rrdtool/rrd_mts_similarity.c

```diff
@@ -0,0 +1,483 @@
+	      rrdms[*num_host].rfs[i].path = strdup(path);
+	      if(rrdms[*num_host].rfs[i].path != NULL)
+	        {
+	            t = true;
+	            (*num_tot_rrds)++;	//counter of all rrds
+	            (num_rrds[i])++;		//counter of each rrd
+	        }
+	      }
+	    }
+	  }
+	}
+      }
+    }
+    free(namelist[n]);
+  }
+  
+  free(namelist);
+
+  /* Report if I've found at least one of the files for a host */
+  if(t)
+    return 1;
+  else
+    return 0;
+}
+
+/* *************************************************** */
+
+int main(int argc, char *argv[]) {
+  rrd_time_value_t start_tv, end_tv;
+  char **filename, *start_s, *end_s, *dirname = NULL, *basedir = NULL;
+  u_int first = 1, quick_mode = 0;
+  float alpha;
+  char c;
+  time_t start, end;
+  u_int num_tot_rrds = 0, i, j = 0;
+  rrd_multifile_stats *rrdms;
+  int n_file = 0, num_host = 0, *num_rrds;
+
+  /* Defaults */
+  alpha   = DEFAULT_ALPHA;
+  start_s = DEFAULT_START;
+  end_s   = DEFAULT_END;
+
+  filename = malloc(sizeof(char*)*MAX_NUM_FILE);
+  for(i=0; i<MAX_NUM_FILE; i++)
+    filename[i] = malloc(sizeof(char)*50);
+
+  while((c = getopt(argc, argv, "d:s:e:a:qf:t:vz")) != '?') {
+    if(c == -1) break;
+
+    switch(c) {
+    case 's':
+      start_s = optarg;
+      break;
+
+    case 'd':
+      basedir = optarg;
+      break;
+
+    case 'e':
+      end_s = optarg;
+      break;
+
+    case 'q':
+      quick_mode = 1;
+      break;
+
+    case 'v':
+      verbose = 1;
+      break;
+
+    case 'a':
+      {
+	float f = atof(optarg);
+
+	if((f > 0) && (f < 1))
+	  alpha = f;
+	else
+	  printf("Discarding -a: valid range is >0 .. <1\n");
+      }
+      break;
+
+    case 'f':
+      filename[n_file] = optarg;
+      n_file++;
+      break;
+
+    case 't':
+      similarity_threshold = atoi(optarg);
+      break;
+
+    case 'z':
+      skip_zero = 1;
+      break;
+
+    default:
+      help();
+      break;
+    }
+  }
+
+  if((filename == NULL) || (basedir == NULL))
+    help();
+
+  if((rrd_parsetime(start_s, &start_tv) != NULL)) {
+    printf("Unable to parse start time %s\n", start_s);
+    return(-1);
+  }
+
+  if((rrd_parsetime(end_s, &end_tv) != NULL)) {
+    printf("Unable to parse end time %s\n", end_s);
+    return(-1);
+  }
+
+  rrd_proc_start_end(&start_tv, &end_tv, &start, &end);
+
+   if((rrdms = ndpi_calloc(sizeof(rrd_multifile_stats), MAX_NUM_USERS)) == NULL) {
+           printf("Not enough memory !\n");
+           return -1;
+   }
+   
+  /* Initializzation of num_rrds */
+  num_rrds = malloc(sizeof(int)*n_file);
+  
+  for(i=0; i<n_file; i++)
+    num_rrds[i] = 0;
+
+  /* Find all rrd's */
+  find_rrds(basedir, filename, rrdms, num_rrds, &num_tot_rrds, n_file, &num_host);
+
+  /* Read RRD's data */
+  for(i=0; i<num_host; i++)
+  { for(j=0; j<n_file; j++)
+    {
+      if(rrdms[i].rfs[j].path != NULL)
+        analyze_rrd(&rrdms[i].rfs[j], start, end);
+    }
+
+    analyze_mts(&rrdms[i], start, end, n_file);
```

Comment:
How to do you specify the RRDs to analzyze? A new command line option is requested as well you need to improve the help